src/coreclr/src/gc/gc.cpp

```diff
@@ -31727,56 +31740,136 @@ void gc_heap::trim_youngest_desired_low_memory()
+    ephemeral_heap_segment->saved_committed = heap_segment_committed(ephemeral_heap_segment);
+    ephemeral_heap_segment->saved_desired_allocation = dd_desired_allocation(dd0);
+#endif // MULTIPLE_HEAPS
+
+#ifndef MULTIPLE_HEAPS
+    // we want to limit the amount of decommit we do per time to indirectly
+    // limit the amount of time spent in recommit and page faults
+    size_t ephemeral_elapsed = dd_time_clock (dd0) - gc_last_ephemeral_decommit_time;
+    gc_last_ephemeral_decommit_time = dd_time_clock (dd0);
 
+    // this is the amount we were planning to decommit
+    ptrdiff_t decommit_size = heap_segment_committed (ephemeral_heap_segment) - decommit_target;
+
+    // we do a max of 160 kB per millisecond (160 MB per second) of pause time
+    // we don't want to take large pause times too seriously - limit to 10 seconds
+    ptrdiff_t max_decommit_size = min(ephemeral_elapsed, 10*1000) * 160 * 1024;
+    decommit_size = min(decommit_size, max_decommit_size);
+
+    slack_space = heap_segment_committed (ephemeral_heap_segment) - heap_segment_allocated (ephemeral_heap_segment) - decommit_size;
     decommit_heap_segment_pages (ephemeral_heap_segment, slack_space);
+#endif // !MULTIPLE_HEAPS
 
     gc_history_per_heap* current_gc_data_per_heap = get_gc_data_per_heap();
     current_gc_data_per_heap->extra_gen0_committed = heap_segment_committed (ephemeral_heap_segment) - heap_segment_allocated (ephemeral_heap_segment);
 }
 
+#ifdef MULTIPLE_HEAPS
+// return true if we actually decommitted anything
+bool gc_heap::decommit_step ()
+{
+    // should never get here for large pages because decommit_ephemeral_segment_pages
+    // will not do anything if use_large_pages_p is true
+    assert(!use_large_pages_p);
+
+    size_t decommit_size = 0;
+    for (int i = 0; i < n_heaps; i++)
+    {
+        gc_heap* hp = gc_heap::g_heaps[i];
+        decommit_size += hp->decommit_ephemeral_segment_pages_step ();
+    }
+    return (decommit_size != 0);
+}
+
+// return the decommitted size
+size_t gc_heap::decommit_ephemeral_segment_pages_step ()
+{
+    // we rely on desired allocation not being changed outside of GC
+    if (ephemeral_heap_segment->saved_desired_allocation != dd_desired_allocation(dynamic_data_of(0)))
+    {
+        GCToOSInterface::DebugBreak();
+    }
+    uint8_t* decommit_target = heap_segment_decommit_target (ephemeral_heap_segment);
+    size_t EXTRA_SPACE = 4 * OS_PAGE_SIZE;
+    decommit_target += EXTRA_SPACE;
+    uint8_t* committed = heap_segment_committed (ephemeral_heap_segment);
+    if (decommit_target < committed)
+    {
+        // we rely on other threads not messing with committed if we are about to trim it down
+        if (ephemeral_heap_segment->saved_committed != heap_segment_committed (ephemeral_heap_segment))
+        {
+            GCToOSInterface::DebugBreak();
+        }
+
+        const size_t DECOMMIT_STEP_SIZE = max (4096 / n_heaps, 100) * OS_PAGE_SIZE;
```

Comment:
instead of doing this calculation here, I would suggest to make 160k a define and use a multiple of that so it's clear how much time you are expressing here for each decommit step. perhaps 4x 160k?

Reply:
I used defines for both the time step in milliseconds and the decommit size per millisecond.